### PR TITLE
Added  more tests to test_whitelist.py

### PR DESCRIPTION
**Pull request by Sekhar-Kumar-Dash**
## Fixes Issue #564 

This pull request adds more tests for whitelist.py

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

### CODE REVIEW

**Comment by Sekhar-Kumar-Dash** [comment]
That's strange. It was working completely when I tested it locally.

**Comment by AlyaGomaa** [comment]
@Sekhar-Kumar-Dash they're failing locally in my computer too!
image

However when i checked your code i noticed the following
image
see the vertical light gray line on the right? this is the line max length. i see your code doesn't follow PEP's max line char per line, you can configure your IDE to display it then i will re-review your PR and we can dbeug what went wrong

File: tests/test_whitelist.py [hunk 1]
```diff
@@ -1,37 +1,561 @@
...
+from slips_files.core.evidence_structure.evidence import (
+    Direction,
+    IoCType 
+    )
+import os
 
+@pytest.fixture
+def mock_db():
+    mock_db = MagicMock()
+    return mock_db
```
[the remark this answers is not in the document]
Comment (Sekhar-Kumar-Dash) [reply tests/test_whitelist.py:16]: I added it when a test was failing because of mockdb 
\

**Comment by Sekhar-Kumar-Dash** [comment]
hey @AlyaGomaa  Just updated this  PR. Actually the problem was that there was a mismatch between my version of the code and the latest version of the code

**Comment by Sekhar-Kumar-Dash** [comment]
hey @AlyaGomaa  it is saying the process is cancelled.
I could not get it did it happen because error in the code

**Comment by AlyaGomaa** [comment]
@Sekhar-Kumar-Dash  i don't know that was weird, but no this is not coming from your code. i restarted CI and we'll see

File: tests/test_whitelist.py [hunk 2]
```diff
@@ -1,37 +1,605 @@
...
+    os.remove(org_info_file)
+
+
+@pytest.mark.parametrize(
+    "mock_ip_info, mock_org_info, ip, org, expected_result", [
+        ({'asn': {'asnorg': 'microsoft'}}, [json.dumps(['microsoft']), json.dumps([])], "91.121.83.118", "microsoft",
+         True),
+        ({'asn': {'asnorg': 'microsoft'}}, [json.dumps(['microsoft']), json.dumps([])], "91.121.83.118", "apple", True),
+        ({'asn': {'asnorg': 'Unknown'}}, json.dumps(['google']), "8.8.8.8", "google", None),
+        ({'asn': {'asnorg': 'AS6432'}}, json.dumps([]), "8.8.8.8", "google", None),
```
Comment (AlyaGomaa) [inline tests/test_whitelist.py:58]: adding comments to your test cases would make it way easier to test and understand what they do, i had to use a debugger to test this one :'D, something like "#testing the case where we dont have ASN info about an org" would be awesome

File: tests/test_whitelist.py [hunk 3]
```diff
@@ -1,37 +1,605 @@
...
+)
+def test_is_ip_asn_in_org_asn(ip, org, org_asn_info, ip_asn_info, expected_result, mock_db):
+    whitelist = ModuleFactory().create_whitelist_obj(mock_db)
+    mock_db.get_org_info.return_value = org_asn_info
+    mock_db.get_ip_info.return_value = ip_asn_info
+    result = whitelist.is_ip_asn_in_org_asn(ip, org)
+    assert result == expected_result
+
+
+def test_parse_whitelist(mock_db):
```
Comment (AlyaGomaa) [inline tests/test_whitelist.py:488]: this function is tested twice

File: tests/test_whitelist.py [hunk 4]
```diff
@@ -1,37 +1,605 @@
...
+
+    org_subnets = whitelist.load_org_IPs('google')
+    assert '34' in org_subnets
+    assert '216' in org_subnets
+    assert '34.64.0.0/10' in org_subnets['34']
+    assert '216.58.192.0/19' in org_subnets['216']
+    os.remove(org_info_file)
+
+
+@pytest.mark.parametrize(
```
Comment (AlyaGomaa) [inline tests/test_whitelist.py:52]: the following format would be more readable
```py
    [
        (
            {"asn": {"asnorg": "microsoft"}},
            [json.dumps(["microsoft"]), json.dumps([])],
            "91.121.83.118",
            "microsoft",
            True,
        ),
        (
            {"asn": {"asnorg": "microsoft"}},
            [json.dumps(["microsoft"]), json.dumps([])],
            "91.121.83.118",
            "apple",
            True,
        ),
        (
            {"asn": {"asnorg": "Unknown"}},
            json.dumps(["google"]),
            "8.8.8.8",
            "google",
            None,
        ),
        (
            {"asn": {"asnorg": "AS6432"}},
            json.dumps([]),
            "8.8.8.8",
            "google",
            None,
        ),
        (
            {"asn": {"asnorg": "google"}},
            json.dumps(["google"]),
            "8.8.8.8",
            "google",
            True,
        ),
        (
            {"asn": {"asnorg": "google"}},
            json.dumps(["google"]),
            "1.1.1.1",
            "cloudflare",
            True,
        ),
        (None, json.dumps(["google"]), "8.8.4.4", "google", None),
    ],
```

File: tests/test_whitelist.py [hunk 5]
```diff
@@ -1,37 +1,605 @@
...
+         {'00:11:22:33:44:55': {'from': 'dst', 'what_to_ignore': 'alerts'}}),
+    ])
+def test_parse_whitelist(whitelist_data, expected_ips, expected_domains, expected_orgs, expected_macs, mock_db):
+    whitelist = ModuleFactory().create_whitelist_obj(mock_db)
+    whitelisted_IPs, whitelisted_domains, whitelisted_orgs, whitelisted_macs = whitelist.parse_whitelist(whitelist_data)
+
+    assert whitelisted_IPs == expected_ips
+    assert whitelisted_domains == expected_domains
+    assert whitelisted_orgs == expected_orgs
+    assert whitelisted_macs == expected_macs
```
Comment (AlyaGomaa) [inline tests/test_whitelist.py:605]: i suggest you run pre-commit hooks on this file, it discovered many issues

File: tests/test_whitelist.py [hunk 6]
```diff
@@ -1,37 +1,605 @@
...
+        # Duplicate entries last one prevails or duplicates included based on implementation
+        ({
+             'IPs': json.dumps({
+                 '1.2.3.4': {'from': 'src', 'what_to_ignore': 'both'},
+                 '1.2.3.4': {'from': 'dst', 'what_to_ignore': 'both'}
+             }),
+             'domains': json.dumps({
+                 'example.com': {'from': 'src', 'what_to_ignore': 'both'},
+                 'example.com': {'from': 'dst', 'what_to_ignore': 'both'}
+             }),
```
Comment (AlyaGomaa) [inline tests/test_whitelist.py:586]: repeated dictionary keys. the second one will override the first in runtime

**Review by AlyaGomaa — COMMENTED** [review]
hey @Sekhar-Kumar-Dash thanks for all the hard work!!
i left some comments, and i will wait for you to fix them and re-review the PR.
again thanks for making Slips better :)!

File: tests/test_whitelist.py [hunk 7]
```diff
@@ -27,11 +28,814 @@ def test_load_org_asn(org, asn, mock_db):
...
+    assert "domains" in all_whitelist
+    assert "organizations" in all_whitelist
+    assert "mac" in all_whitelist
+
+
+@pytest.mark.parametrize(
+    "flow_data, whitelist_data, expected_result",
+    [
+        (  # testing_is_whitelisted_flow_with_whitelisted_organization_but_ip_or_domain_not_whitelisted
+            MagicMock(
```
Comment (AlyaGomaa) [inline tests/test_whitelist.py:771]: hey you can use spaces as word separators in comments instead of _ :))

File: tests/test_whitelist.py [hunk 8]
```diff
@@ -9,12 +13,9 @@ def test_read_whitelist(mock_db):
...
     mock_db.get_whitelist.return_value = {}
-    (
-        whitelisted_IPs,
-        whitelisted_domains,
-        whitelisted_orgs,
-        whitelisted_mac,
-    ) = whitelist.read_whitelist()
+    whitelisted_IPs, whitelisted_domains, whitelisted_orgs, whitelisted_mac = (
+        whitelist.read_whitelist()
+    )
```
Comment (AlyaGomaa) [inline tests/test_whitelist.py:18]: i think we can use mock file for testing the read_whitelist function, this is an important one so we probably should have more strict tests for it. for example 2  whitelist test files would be enough, and in these files we can check for expected values, missing columns, missing values in a line, etc. what do you think @Sekhar-Kumar-Dash

File: tests/test_whitelist.py [hunk 9]
```diff
@@ -27,11 +28,814 @@ def test_load_org_asn(org, asn, mock_db):
...
+                        "what_to_ignore": "flows",
+                    }
+                }
+            },
+            False,
+        ),
+    ],
+)
+def test_is_whitelisted_flow(
+    mock_db, flow_data, whitelist_data, expected_result
```
Comment (AlyaGomaa) [inline tests/test_whitelist.py:834]: i will split  is_whitelisted_flow() into smaller functions and will let you know so we can better test it, it's very long now and this test case is testing too many different things at once

**Review by AlyaGomaa — COMMENTED** [review]
Hello @Sekhar-Kumar-Dash i added some comments, please don't start resolving them until i push my refactoring for whitelist.py so we can better test large functions. thanksssss

File: tests/test_whitelist.py (hunk 8 above)
Comment (Sekhar-Kumar-Dash) [reply tests/test_whitelist.py:18]: yes you are right

File: tests/test_whitelist.py (hunk 7 above)
Comment (Sekhar-Kumar-Dash) [reply tests/test_whitelist.py:771]: That's just a mistake I made while typing quickly. Sorry about that.

File: tests/test_whitelist.py [hunk 10]
```diff
@@ -27,11 +28,814 @@ def test_load_org_asn(org, asn, mock_db):
...
+    mock_db.get_all_whitelist.return_value = {
+        "IPs": json.dumps(
+            {"1.2.3.4": {"from": "src", "what_to_ignore": "both"}}
+        ),
+        "domains": json.dumps(
+            {"example.com": {"from": "dst", "what_to_ignore": "both"}}
+        ),
+    }
+    mock_db.is_whitelisted_tranco_domain.return_value = False
+    assert whitelist.is_whitelisted_victim(victim_data) == expected_result
```
[the remark this answers is not in the document]
Comment (Sekhar-Kumar-Dash) [reply tests/test_whitelist.py:586]: I think the there is something wrong with the actual code let's check it

**Pull request merged**